File: Data Structures & Algorithms/4sum/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[][]}
     */
    fourSum(nums, target) {
        nums.sort((a,b)=>a-b)
        let result = []
        
        for(let i=0;i<nums.length;i++){
            if(i>0 && nums[i]===nums[i-1]) continue
            for(let j=i+1;j<nums.length;j++){
                if(j>i+1 && nums[j]===nums[j-1]) continue
                let start=j+1, end=nums.length-1
                while(start<end){
                    let sum = nums[i]+nums[j]+nums[start]+nums[end]
                    if(sum>target){
                        end--
                    } else if(sum<target){
                        start++
                    } else{
                        result.push([nums[i],nums[j],nums[start],nums[end]])
                        start++
                        while(start<end && nums[start]===nums[start-1]){
                            start++
                        }
                    }
                }
            }
        }
        return result
    }
}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res, subset = [], []

        def dfs(start, cur):
            if cur==target:
                res.append(subset[:])
                return
            if cur>target:
                return
            
            for i in range(start, len(candidates)):
                if i>start and candidates[i]==candidates[i-1]:
                    continue
                
                subset.append(candidates[i])
                dfs(i+1, cur+candidates[i])
                subset.pop()
            
        dfs(0,0)
        return res